plugin.js

```diff
@@ -51,7 +51,9 @@ function fastifyCookieClearCookie (reply, name, options) {
 }
 
 function onReqHandlerWrapper (fastify) {
-  return function fastifyCookieOnReqHandler (fastifyReq, fastifyRes, done) {
+  return function fastifyCookieHandler (...argv) {
+    const fastifyReq = argv[0]
+    const done = argv.pop()
```

Comment:
Why this change?

Reply:
on some hooks, the done function is the fourth argument

Reply:
But we know the arity of the hooks. Also spreading arguments is converting the iterable arguments into an array.

So we can also pass onReqHandlerWrapper the hook as second parameter and then do

```
function onReqHandlerWrapper (fastify, hook) {
  const donePosition = hook === 'preParsing' ? 3 : 2
  return function fastifyCookieHandler () {
    const fastifyReq = arguments[0]
    const done = arguments[donePosition]
    // [...]
```
no popping needed. 

Or like this:

```
function onReqHandlerWrapper (fastify, hook) {
  return hook === 'preParsing'
    ? function fastifyCookieHandler (fastifyReq, fastifyRes, payload, done) {
        fastifyReq.cookies = {} // New container per request. Issue #53
        const cookieHeader = fastifyReq.raw.headers.cookie
        if (cookieHeader) {
          fastifyReq.cookies = fastify.parseCookie(cookieHeader)
        }
        done()
      }
    : function fastifyCookieHandler (fastifyReq, fastifyRes, done) {
      fastifyReq.cookies = {} // New container per request. Issue #53
      const cookieHeader = fastifyReq.raw.headers.cookie
      if (cookieHeader) {
        fastifyReq.cookies = fastify.parseCookie(cookieHeader)
      }
      done()
    }
}
```

Reply:
Or

```
function onReqHandlerWrapper (fastify, hook) {
  function hookHandler (fastifyReq, done) {
    fastifyReq.cookies = {} // New container per request. Issue #53
    const cookieHeader = fastifyReq.raw.headers.cookie
    if (cookieHeader) {
      fastifyReq.cookies = fastify.parseCookie(cookieHeader)
    }
    done()
  }

  return hook === 'preParsing'
    ? function fastifyCookieHandler (fastifyReq, fastifyRes, payload, done) {
        hookHandler(fastifyReq, done)
      }
    : function fastifyCookieHandler (fastifyReq, fastifyRes, done) {
      hookHandler(fastifyReq, done)
    }
}

```

Reply:
I mean... creating an array and popping the last document etc. is probably not that performant :(

Reply:
and if there are other hooks, we always know that NextFunction comes last.  

pop() has high overhead?

Reply:
maybe so?

```
const fastifyReq = argv[0]
const done = argv[argv.length-1]
```

Reply:
creating an array to just pop the last element increases the overhead. Is it low? Is it high? Idk, but why should we not choose the implementation, which has the lowest overhead of them all. And one were i create less intermediary objects, is the better one (memory management)

And to be honest I prefer the second solution, because I know that it has the lowest overhead of all the impementations.

And yes we support currently only 4 hooks. And if we have to support one more, we can add it and check the arity. 

Tbh, when you created the Object for the getHook-check you wrote that it has faster access. I thought ok, if he wants that and it is faster no problem, despite the fact, that getHook will be called by a service only one time - at the startup. So the speed is not that relevant but faster startup is always desirable.

We talk here about a hook function. This will run on every request. So I expect a slow down with the proposed implementation.

Reply:
Didnt see that you wrote the question, while I was writing my comment.

the issue is that you always do `...argv` which is basically transforming the arguments iterable to an array. Then you do a argv.length, so you have to lookup the length and finally subtract by 1. Also you need  instantiate two variables.

And this will be done for every request, despite the fact, that argv.length will always be the same. 

I dont know what the others think about it ( @mcollina , @Eomm @climba03003 ) . Maybe wait for some feedback.

Reply:
I agree, using a prepared handler will be faster than working with an array in every request. I'm for performance :) 

In the second solution, there is code repetition. It doesn’t look very good, but if you know for sure that it’s faster this way, then I’ll change the solution to this.